File: src/learning_service/src/add_client.cpp
#include <ros/ros.h>
#include <learning_service/Add.h>
#include <cstdlib>

int main(int argc, char **argv)
{
    ros::init(argc, argv, "add_client");

    // 判断在客户端输入的string型参数的数量是否为3个, 因为argc为封装的参数个数+1,
    // 所以为了限定输入参数为2, 我们将其限定值为3, 如果其值不为3, 系统会自动报错.
    if (argc != 3)
    {
        ROS_ERROR("usage: add two number X Y");
        return 1;
    }

    // 等待服务端创建的服务对象, 为之后的客户端发送服务请求做准备.
    // 只有寻找到对应的服务时, 客户端才会发送对应的服务请求.
    ros::service::waitForService("add_service");

    ros::NodeHandle nh;
    // 通过句柄nh, 将服务类型定义为我们自定义的learning_service::Add型, 将其连接到服务add_service.
    ros::ServiceClient add_client = nh.serviceClient<learning_service::Add>("add_service");

    // 初始化自定义服务中request内容, 其数据源于数组argv中封装的内容, 但因为初始化中的输入参数为string型变量,
    // 与我们在srv中定义的float64型的变量不一致, 所以需要先将其类型做转换后再对srv中的request变量进行赋值.
    // 其中atof()函数用于将string型变量转换为我们需要的float型变量, argv[1]中存放的输入的第一个参数,
    // argv[2]存放的是输入的第二个参数.
    learning_service::Add srv;
    srv.request.a = atof(argv[1]);
    srv.request.b = atof(argv[2]);

    // 调用与客户端相连接的服务, 并判断调用是否成功, 如果成功则在终端中显示服务调用的response, 如果调用失败同样
    // 在终端中发出调用失败的警告, 其中, add_client.call(srv)如果执行成功会返回数值1, 可以用于判断调用成功与否.
    if (add_client.call(srv))
    {
        ROS_INFO("sum is: %0.2f", srv.response.sum);
    }
    else
    {
        ROS_ERROR("Failed to call service add_service");
        return 1;
    }
    return 0;
}

File: src/learning_service/scripts/add_client.py
#!/usr/bin/env python3
# coding:utf-8

import sys

import rospy

from learning_service.srv import Add  # 包含自定义的Add.srv的服务类型, 用于后续传送服务的数据.


def add_client(x, y):
    # 初始化一个名为add_client的节点, 并等待由客户端创建的服务add_service, 为之后客户端向该服务发送数据做准备.
    rospy.init_node("add_client", anonymous=True)  # 注意这里节点名是 add_client
    rospy.wait_for_service("add_service")  # 注意这里服务是 add_service

    try:  # 构造客户端实例, 通过客户端将从终端中获取的参数x, y传递给服务从而获得服务的反馈值.
        # 生成一个名为c的客户端实例, 定义其服务类型为Add型, 将其连接到指定服务add_service.
        c = rospy.ServiceProxy("add_service", Add)
        # 用于调用服务, 即将终端中得到的参数传递到服务中, 并将该服务中的数据传递给变量response, 用于获取服务反馈值.
        response = c(x, y)
        rospy.loginfo("Sum is: %s", response.sum)
    except rospy.ServiceException as e:
        rospy.loginfo("Failed to call service add_service: %s", e)


if __name__ == "__main__":
    if len(sys.argv) == 3:  # 接收终端中输入的参数, 并将参数赋值为x和y, 将其作为参数传递给add_client函数.
        x = float(sys.argv[1])
        y = float(sys.argv[2])
        add_client(x, y)
    else:
        rospy.logerr("usage: add two ints X Y")
        sys.exit(1)
